File: runtimes/compose/src/main/java/com/styleconverter/runtime/typography/font/DocumentFontRegistry.kt
package com.styleconverter.runtime.typography.font

import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import com.styleconverter.runtime.core.ir.IRFontFace
import java.io.File

/**
 * DocumentFontRegistry — runtime registration of the document's `@font-face`
 * declarations (schema/spec/01-envelope.md §5; wave-35 lane B2).
 *
 * ## Why this exists, and why it is not a property triplet
 *
 * Everything under this runtime's category folders is a per-PROPERTY triplet:
 * one IR property in, one Compose modifier out. A font FACE is neither. It is
 * a document-level RESOURCE that must be registered with the text stack before
 * any element referencing it measures, and registering it means knowing where
 * the HOST put the file. So the registry is a document-scoped object the host
 * fills in (the harness's inbox path does it right after
 * `IRDocumentDecoder.decode`), and the per-property resolver
 * [com.styleconverter.runtime.typography.CssFontFamilyResolver] merely
 * CONSULTS it — exactly the split the web side uses between
 * `apps/web-harness/src/sdui/useFontFaces.ts` (mount) and the engine (read).
 *
 * ## What it closes
 *
 * Until wave 35 both natives DECODED `fontFaces` and ignored it: a test
 * declaring `@font-face { font-family: test; src: url(X.woff) }` plus
 * `body { font: 36px test }` shaped with the bundled Inter while the browser
 * ref painted the author's outlines. css-text/boundary-shaping-001…008 are the
 * measured case — they assert on "fi"/"ffi" LIGATURES that only the declared
 * LinLibertine face carries, so the assertion was unobservable on the native
 * side of the diff (wpt-not-applicable.mjs Rule 15's closing move, step 1).
 *
 * ## The file channel
 *
 * The wire carries a corpus-relative PATH, never a payload. The feeder
 * (tools/titan/feed-android.mjs) copies each referenced file into the app's
 * own sandbox at `<externalFilesDir>/fonts/<src>`, preserving the relative
 * path VERBATIM — so resolution here is a plain [File] join with no name
 * mangling and no escaping rule that could drift from the host's.
 *
 * ## Degradation is loud, never silent
 *
 * A face whose file is absent, unreadable or unparseable is DECLINED: the
 * family is simply not registered, [resolve] answers null, and the existing
 * resolver walk continues to the fallback exactly as it did in wave 34. Each
 * decline is logged at WARN and counted in [lastReport], because the failure
 * mode this guards against — rendering plausible text in the WRONG face — is
 * invisible in a screenshot.
 */
object DocumentFontRegistry {

    private const val TAG = "DocFontRegistry"

    /** Logging seam. `runCatching` guards `android.util.Log` for JVM callers —
     *  the module's unit suite runs without `returnDefaultValues`, so an
     *  unguarded Log call throws "not mocked" and would make this whole
     *  registry untestable off-device. Same guard InlineRunPlan uses. */
    private fun logw(message: String, e: Throwable? = null) {
        runCatching { android.util.Log.w(TAG, message, e) }
    }

    private fun logi(message: String) {
        runCatching { android.util.Log.i(TAG, message) }
    }

    /** How a readable file becomes a Compose [Font].
     *
     *  A SEAM, and it exists for one measured reason:
     *  `androidx.compose.ui.text.font.Font(File, …)` calls
     *  `android.graphics.Typeface.createFromFile` EAGERLY, and that method is
     *  not mocked in a JVM unit test — so with the call inlined, every
     *  registration in the unit suite fell into the decline branch and the
     *  grouping/matching rules below could not be pinned at all. Swapping the
     *  loader lets the suite exercise everything this file OWNS (family
     *  identity, §4.1 grouping, decline handling, the resolver hook) while the
     *  device gate measures the one thing it does not own — whether Android
     *  can rasterise the bytes.
     *
     *  `internal` + restored by the test's tearDown: production has exactly
     *  one implementation and no way to reach this from outside the module. */
    internal var fontLoader: (File, FontWeight, FontStyle) -> Font =
        { file, weight, style -> Font(file = file, weight = weight, style = style) }

    /** Font container formats `android.graphics.Typeface.createFromFile` can
     *  actually parse — and, critically, NOT the web-only ones.
     *
     *  MEASURED (wave-35 device gate, emulator API 36, the css-text
     *  boundary-shaping face `LinLibertine_Re-4.7.5.woff`): handing a WOFF to
     *  `Font(file = …)` neither throws nor loads. `Typeface.createFromFile`
     *  answers the DEFAULT typeface, so registration "succeeds", the family
     *  resolves, and the capture renders in the default face — byte-identical
     *  to the no-face control (0 differing pixels of 234 000). That is the
     *  precise failure this whole channel exists to end: a plausible-looking
     *  screenshot shaped from the wrong outlines, with nothing in any log.
     *  The identical document with a TTF face (`fonts/Ahem.ttf`) moves ink
     *  0.597% → 6.092% (13 837 pixels), which is what proves the resolution
     *  path itself is correct and the gap is FORMAT, not plumbing.
     *
     *  iOS is deliberately NOT symmetric here: CoreText's
     *  `CTFontManagerRegisterFontsForURL` DOES accept WOFF (same device gate:
     *  the same LinLibertine face moved iOS ink 0.665% → 0.488%), so its
     *  registry admits everything the wire admits. Mirroring this table there
     *  would decline a face that platform can genuinely render.
     *
     *  So: decline LOUDLY here rather than pretend. A declined face falls back
     *  to the resolver walk — the wave-34 behaviour — and says so. Closing the
     *  gap for real means transcoding WOFF → TTF in the delivery hop
     *  (tools/titan/feed-android.mjs), which is a follow-up, not a silent
     *  fallthrough. */
    private val ANDROID_LOADABLE_EXTENSIONS = setOf("ttf", "otf", "ttc")

    /** Registered families, keyed by the css-fonts-4 §4.2 ASCII-case-
     *  insensitive family name. Replaced wholesale by [register] rather than
     *  mutated, so a reader that grabbed the map mid-swap sees a consistent
     *  generation instead of a half-built one (the harness registers on the
     *  main thread and Compose reads during composition — same thread today,
     *  but the volatile swap costs nothing and removes the assumption). */
    @Volatile
    private var families: Map<String, FontFamily> = emptyMap()

    /** Outcome of the most recent [register] call. Exposed so the harness can
     *  log a per-document line and a test can assert delivery without
     *  reaching into Compose's font stack. */
    data class Report(
        val declared: Int = 0,
        val registered: Int = 0,
        val declined: List<String> = emptyList()
    )

    @Volatile
    var lastReport: Report = Report()
        private set

    /**
     * Register a document's faces, replacing any previous document's.
     *
     * REPLACE, not merge: the inbox harness renders many documents in one
     * process, and a face left over from document N-1 would shadow document
     * N's same-named family — the "looks like text, shaped from the wrong
     * file" failure this whole channel exists to end. The feeders wipe the
     * on-device fonts directory per run for the same reason.
     *
     * @param faces the decoded `IRDocument.fontFaces` (null/empty ⇒ clear)
     * @param baseDir the directory the host copied the font FILES under; the
     *   entry's `src` is joined onto it verbatim. null ⇒ nothing can resolve,
     *   so every declared face is declined (and logged).
     * @return the number of FAMILIES registered
     */
    fun register(faces: List<IRFontFace>?, baseDir: File?): Int {
        if (faces.isNullOrEmpty()) {
            families = emptyMap()
            lastReport = Report()
            return 0
        }
        // Group by family first: css-fonts-4 §4.1 lets one family be declared
        // by SEVERAL faces (regular + bold + italic files), and Compose models
        // that as ONE FontFamily holding several Fonts. Building one family
        // per entry would make the last entry win and drop the other weights.
        val byFamily = LinkedHashMap<String, MutableList<Font>>()
        val declined = mutableListOf<String>()
        for (face in faces) {
            val file = baseDir?.let { File(it, face.src) }
            if (file == null || !file.isFile || !file.canRead()) {
                // No silent fallthrough: name the family AND the path so an
                // investigator can tell a missing hop from a bad path.
                logw("declined @font-face '${face.family}': no readable file at " +
                        "${file?.absolutePath ?: "<no baseDir>"} — falling back to the resolver walk")
                declined += face.family
                continue
            }
            // Format gate BEFORE the load: see ANDROID_LOADABLE_EXTENSIONS.
            // Checked here rather than inside the try because a silent default
            // is not an exception — there is nothing to catch.
            val ext = file.name.substringAfterLast('.', "").lowercase()
            if (ext !in ANDROID_LOADABLE_EXTENSIONS) {
                logw("declined @font-face '${face.family}': ${file.name} is a '$ext' " +
                        "container, which android.graphics.Typeface cannot parse — it would " +
                        "load as the DEFAULT typeface and silently render the wrong outlines")
                declined += face.family
                continue
            }
            val loaded = try {
                // androidx.compose.ui.text.font.Font(File, …) — the Compose
                // API for a face that is not an app resource. Weight/style
                // come from the DESCRIPTORS, not from the file's own metadata,
                // because §4.4/§4.5 make the descriptor authoritative for
                // matching (a "bold" descriptor on a regular file must still
                // answer bold requests).
                fontLoader(file, parseWeight(face.weight), parseStyle(face.style))
            } catch (e: Exception) {
                // Compose throws when the file is not a font Android can parse
                // (Typeface.createFromFile rejects it).
                logw("declined @font-face '${face.family}': ${file.name} did not load", e)
                declined += face.family
                continue
            }
            byFamily.getOrPut(familyKey(face.family)) { mutableListOf() } += loaded
        }
        val built = byFamily.mapValues { (_, fonts) -> FontFamily(fonts) }
        families = built
        lastReport = Report(declared = faces.size, registered = built.size, declined = declined)
        if (built.isNotEmpty()) {
            logi("registered ${built.size} @font-face famil${if (built.size == 1) "y" else "ies"}: " +
                    built.keys.joinToString(", "))
        }
        return built.size
    }

    /** Forget every registered face. Used by tests and by a host that decodes
     *  a document with no `fontFaces` (the harness calls [register] with null,
     *  which routes here) so nothing leaks across documents. */
    fun clear() {
        families = emptyMap()
        lastReport = Report()
    }

    /**
     * The registered family for one CSS family name, or null when this
     * document declared no such face.
     *
     * Null is the contract, not a fallback: the caller
     * ([com.styleconverter.runtime.typography.CssFontFamilyResolver]) is
     * walking a css-fonts-4 §5.2 fallback list and must be free to continue to
     * the next name. Answering a default here would stop the walk at the first
     * name and re-introduce exactly the wrong-face bug.
     */
    fun resolve(family: String): FontFamily? = families[familyKey(family)]

    /** True when this document registered at least one face — the cheap
     *  guard the resolver uses to stay byte-for-byte on its wave-34 path for
     *  the (overwhelmingly common) face-free document. */
    fun isEmpty(): Boolean = families.isEmpty()

    /** css-fonts-4 §4.2 family identity: ASCII case-insensitive, quotes
     *  stripped. Mirrors CssFontFamilyResolver.resolveEntry's own
     *  normalisation so a name that matches there matches here. */
    private fun familyKey(raw: String): String =
        raw.trim().trim('"', '\'').lowercase()

    /** The §4.4 `font-weight` descriptor → Compose [FontWeight].
     *
     *  The descriptor is carried AS AUTHORED (it may be a RANGE like
     *  "400 700"), so this reads the FIRST numeric token: Compose has no
     *  variable-range font family API at this BOM, and the range's low end is
     *  the face's own natural weight. Keywords map per §4.4; anything
     *  unrecognised falls to Normal, which is the §4.4 initial. */
    internal fun parseWeight(descriptor: String?): FontWeight {
        val d = descriptor?.trim()?.lowercase() ?: return FontWeight.Normal
        if (d.isEmpty()) return FontWeight.Normal
        when (d) {
            "normal" -> return FontWeight.Normal
            "bold" -> return FontWeight.Bold
        }
        val first = d.split(Regex("\\s+")).firstOrNull()?.toIntOrNull() ?: return FontWeight.Normal
        // §4.4 clamps the descriptor to [1, 1000]; Compose's FontWeight is a
        // plain int wrapper, so clamp rather than construct an invalid one.
        return FontWeight(first.coerceIn(1, 1000))
    }

    /** The §4.5 `font-style` descriptor → Compose [FontStyle].
     *
     *  `oblique` (with or without an angle) maps to Italic: Compose exposes
     *  only the two-valued FontStyle, and §4.5 makes oblique the synthesised-
     *  slant sibling of italic. The angle is deliberately dropped rather than
     *  approximated — a wrong slant is a visible lie, a missing one is the
     *  documented platform limit. */
    internal fun parseStyle(descriptor: String?): FontStyle {
        val d = descriptor?.trim()?.lowercase() ?: return FontStyle.Normal
        return if (d.startsWith("italic") || d.startsWith("oblique")) FontStyle.Italic
        else FontStyle.Normal
    }
}

File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/typography/font/DocumentFontRegistry.swift
//
//  DocumentFontRegistry.swift
//  StyleConverterRuntime — wave-35 lane B2
//
//  Runtime registration of the document's `@font-face` declarations
//  (schema/spec/01-envelope.md §5).
//
//  ## Why this exists, and why it is not a property triplet
//
//  Everything else in this folder is a per-PROPERTY triplet: one IR property
//  in, one SwiftUI modifier out. A font FACE is neither — it is a
//  document-level RESOURCE that must be registered with CoreText BEFORE any
//  Text measures, and registering it means knowing where the HOST put the
//  file. So the registry is a document-scoped object the host fills in (the
//  harness's inbox path calls `register` right after the IRDocument decode)
//  and the per-property path merely CONSULTS it. Exactly the split the web
//  side uses between apps/web-harness/src/sdui/useFontFaces.ts (mount) and the
//  engine (read), and the Compose side uses between its own
//  DocumentFontRegistry and CssFontFamilyResolver.
//
//  ## What it closes
//
//  Until wave 35 both natives DECODED `fontFaces` and ignored it (IRModels'
//  own comment said so): a test declaring
//  `@font-face { font-family: test; src: url(X.woff) }` plus
//  `body { font: 36px test }` shaped with the bundled Inter while the browser
//  ref painted the author's outlines. css-text/boundary-shaping-001…008 are
//  the measured case — they assert on "fi"/"ffi" LIGATURES only the declared
//  LinLibertine face carries.
//
//  ## The CoreText hop, and the NAME problem it creates
//
//  `CTFontManagerRegisterFontsForURL` makes the file's own faces available to
//  the process — under the FILE's family name ("Linux Libertine"), not the
//  CSS family name ("test"). SwiftUI's `.custom(_:size:)` and `UIFont(name:)`
//  both take the platform name, so registration alone changes nothing: the
//  CSS name still resolves to nothing and falls back. That is why this file
//  keeps a MAPPING — CSS family (case-insensitively) → the PostScript name
//  CoreText reported for the file it just registered — and why the mapping is
//  read off the registered descriptors rather than guessed from the filename.
//
//  ## Degradation is loud, never silent
//
//  A face whose file is missing, unreadable or unregisterable is DECLINED: no
//  mapping is added, `resolvedName` answers nil, and the existing fallback
//  walk runs exactly as it did in wave 34. Every decline is logged with the
//  family AND the path, because the failure this guards against — plausible
//  text in the WRONG face — is invisible in a screenshot.
//

import Foundation
import CoreText
#if canImport(UIKit)
import UIKit
#endif

/// Process-wide registry of the current document's `@font-face` faces.
public final class DocumentFontRegistry {

    /// The one instance the harness fills and the font path reads. A
    /// singleton because CoreText registration is itself process-scoped:
    /// a second registry could not undo the first's registrations, so
    /// pretending they are independent would be a lie about the platform.
    public static let shared = DocumentFontRegistry()

    /// Outcome of the most recent `register` call — surfaced so the harness
    /// can log one line per document and a test can assert delivery without
    /// reaching into CoreText.
    public struct Report: Equatable {
        public let declared: Int
        public let registered: Int
        public let declined: [String]
        public init(declared: Int = 0, registered: Int = 0, declined: [String] = []) {
            self.declared = declared
            self.registered = registered
            self.declined = declined
        }
    }

    /// CSS family (lowercased, unquoted) → the platform font name to hand to
    /// `.custom(_:size:)`.
    private var mapping: [String: String] = [:]

    /// URLs this registry handed to CoreText, so `clear()` can hand them back.
    /// Without this a face from document N-1 stays registered for the whole
    /// process and shadows document N's same-named family — "looks like text,
    /// shaped from the wrong file", the exact failure this channel ends.
    private var registeredURLs: [URL] = []

    public private(set) var lastReport = Report()

    /// wave-47 lane Z4 — registration epoch: bumped by every [clear] (and so
    /// by every [register], which clears first). ChUnitMetrics folds it into
    /// its memoized-advance key so a '0' advance cached for document N's face
    /// can never answer for document N+1 — the registry replaces its mapping
    /// wholesale per document, and two documents may bind one PostScript name
    /// to different files.
    public private(set) var epoch: Int = 0

    private init() {}

    /// css-fonts-4 §4.2 family identity: ASCII case-insensitive, quotes
    /// stripped. Mirrors the Compose registry's `familyKey` and the web
    /// harness's family handling so a name that matches on one surface
    /// matches on all three.
    private static func familyKey(_ raw: String) -> String {
        raw.trimmingCharacters(in: .whitespacesAndNewlines)
            .trimmingCharacters(in: CharacterSet(charactersIn: "\"'"))
            .lowercased()
    }

    /// Register a document's faces, REPLACING any previous document's.
    ///
    /// - Parameters:
    ///   - faces: decoded `IRDocument.fontFaces` (nil/empty ⇒ clear only).
    ///   - baseDirectory: the directory the host copied the font FILES under.
    ///     The entry's `src` is appended verbatim (the feeder preserves the
    ///     corpus-relative path exactly so there is no escaping rule to drift
    ///     between the four codebases). nil ⇒ every face is declined.
    /// - Returns: the number of CSS families that now resolve.
    @discardableResult
    public func register(_ faces: [IRFontFace]?, baseDirectory: URL?) -> Int {
        clear()
        guard let faces, !faces.isEmpty else { return 0 }
        var declined: [String] = []
        for face in faces {
            guard let base = baseDirectory else {
                log("declined @font-face '\(face.family)': no base directory for \(face.src)")
                declined.append(face.family)
                continue
            }
            // `appendingPathComponent` on a multi-segment relative path keeps
            // the segments — which is what the verbatim contract needs.
            let url = base.appendingPathComponent(face.src)
            guard FileManager.default.fileExists(atPath: url.path) else {
                log("declined @font-face '\(face.family)': no file at \(url.path)")
                declined.append(face.family)
                continue
            }
            // Read the file's own descriptors BEFORE registering: a file
            // CoreText cannot describe is one it cannot register either, and
            // failing here costs nothing to undo.
            guard let platformName = Self.platformFontName(of: url) else {
                log("declined @font-face '\(face.family)': CoreText could not read a font name from \(url.lastPathComponent)")
                declined.append(face.family)
                continue
            }
            var error: Unmanaged<CFError>?
            // `.process` scope: visible to this app only, and — crucially —
            // UNREGISTERABLE later, which `clear()` depends on.
            let ok = CTFontManagerRegisterFontsForURL(url as CFURL, .process, &error)
            if !ok {
                // Already-registered is the one benign failure: the same file
                // fed twice in one process (two documents, one face) is fine
                // and the mapping below is still correct.
                let code = (error?.takeRetainedValue()).map { CFErrorGetCode($0) }
                if code != CTFontManagerError.alreadyRegistered.rawValue {
                    log("declined @font-face '\(face.family)': registration failed for \(url.lastPathComponent) (code \(code.map(String.init) ?? "?"))")
                    declined.append(face.family)
                    continue
                }
            } else {
                registeredURLs.append(url)
            }
            // Last declaration of a family wins — css-fonts-4 §4.1's rule for
            // two faces occupying the same slot, and the same rule the section
            // pipeline's combined-fixture merge enforces upstream.
            mapping[Self.familyKey(face.family)] = platformName
        }
        lastReport = Report(declared: faces.count, registered: mapping.count, declined: declined)
        if !mapping.isEmpty {
            log("registered \(mapping.count) @font-face famil\(mapping.count == 1 ? "y" : "ies"): " +
                mapping.map { "\($0.key)→\($0.value)" }.sorted().joined(separator: ", "))
        }
        return mapping.count
    }

    /// The platform font name for a CSS family, or nil when this document
    /// declared no such face.
    ///
    /// nil is the contract, not a fallback: the caller is walking a
    /// css-fonts-4 §5.2 fallback list and must stay free to continue to the
    /// next name. Answering a default here would stop the walk at the first
    /// name and re-introduce the wrong-face bug this file exists to end.
    public func resolvedName(for cssFamily: String) -> String? {
        mapping[Self.familyKey(cssFamily)]
    }

    /// True when this document registered no face — the cheap guard the font
    /// path uses to stay byte-for-byte on its wave-34 behaviour for the
    /// (overwhelmingly common) face-free document.
    public var isEmpty: Bool { mapping.isEmpty }

    /// Unregister everything and forget the mapping.
    public func clear() {
        for url in registeredURLs {
            var error: Unmanaged<CFError>?
            // Best-effort: a face still referenced by a live layout can refuse
            // to unregister. The mapping is dropped either way, so the stale
            // face can no longer be NAMED even if it lingers in CoreText.
            _ = CTFontManagerUnregisterFontsForURL(url as CFURL, .process, &error)
        }
        registeredURLs.removeAll()
        mapping.removeAll()
        lastReport = Report()
        // The document boundary: everything memoized against the old faces
        // (ChUnitMetrics' advance cache) expires with them — see `epoch`.
        epoch += 1
    }

    /// The name `.custom(_:size:)` / `UIFont(name:)` will accept for the first
    /// face in a font file, read from CoreText's own descriptors.
    ///
    /// PostScript name, not family name: `UIFont(name:)` accepts either, but
    /// the PostScript name is unique per FACE, so a file carrying several
    /// faces cannot resolve to an ambiguous family string.
    static func platformFontName(of url: URL) -> String? {
        guard let descriptors = CTFontManagerCreateFontDescriptorsFromURL(url as CFURL) as? [CTFontDescriptor],
              let first = descriptors.first else { return nil }
        let name = CTFontDescriptorCopyAttribute(first, kCTFontNameAttribute) as? String
        if let name, !name.isEmpty { return name }
        // Fall back to the family attribute: a few older faces omit the
        // PostScript name. Still a real name from the file, never a guess.
        let family = CTFontDescriptorCopyAttribute(first, kCTFontFamilyNameAttribute) as? String
        return (family?.isEmpty == false) ? family : nil
    }

    /// Single logging seam. `print` matches the harness's other TITAN
    /// diagnostics (they are scraped out of the simulator log), and keeping it
    /// in one place means a future switch to os_log touches one line.
    private func log(_ message: String) {
        print("[StyleConverter] DocumentFontRegistry: \(message)")
    }
}
